numpy/_core/src/umath/stringdtype_ufuncs.cpp

```diff
@@ -1028,6 +1028,25 @@ all_strings_promoter(PyObject *NPY_UNUSED(ufunc),
                      PyArray_DTypeMeta *const signature[],
                      PyArray_DTypeMeta *new_op_dtypes[])
 {
+    if (op_dtypes[0] != &PyArray_StringDType &&
+            op_dtypes[1] != &PyArray_StringDType &&
+            op_dtypes[2] != &PyArray_StringDType) {
+        /*
+         * This promoter was triggered with only unicode arguments, so use
+         * unicode.  This can happen due to `dtype=` support which sets the
+         * output DType/signature.
+         */
+        new_op_dtypes[0] = NPY_DT_NewRef(&PyArray_UnicodeDType);
+        new_op_dtypes[1] = NPY_DT_NewRef(&PyArray_UnicodeDType);
+        new_op_dtypes[2] = NPY_DT_NewRef(&PyArray_UnicodeDType);
+        return 0;
+    }
+    if (signature[0] == &PyArray_UnicodeDType &&
+            signature[1] == &PyArray_UnicodeDType &&
+            signature[2] == &PyArray_UnicodeDType) {
+        /* Unicode forced, but didn't override a string input: invalid */
+        return -1;
```

Comment:
This part makes me wonder if I should just check it after the promoter is done and invalidate the result if this is violated.  But it is OK here also.

Reply:
I agree it would be better to enforce that there, if only because IMO DType authors shouldn't have to worry about that case or add code to account for it to write a correct DType.

Reply:
Indeed, it seems strange one would even get here if the signature is already clear that `StringDType` should not be involved.